sqlccl: specify temp directory during local CSV transforms

madelynnblue opened the pull request:
Change the local CSV transform to take a RocksDB engine instead of
a directory, which allows for correctly supporting the CLI (OS temp)
and server (temp storage) use cases.

This adds a new flag to the `load csv` subcommand: tempdir, which
overrides the OS temp dir used for intermediate storage. Notably the
default is listed in the help text, making it explicit to the user
where the data will be stored if no option is passed.

For the SQL invocation that uses the local transform, pass down the
server's temp storage instead of using the OS temp dir.

This adds a new function that creates a RocksDB store instead of a
map. That is, it allows for duplicate keys instead of overwriting
them. This functionality is needed for CSV writing, which requires
that duplicate keys are kept so it can detect primary key or unique
index violations. The CSV code was previously doing that itself, but
has been moved to use the common temp storage location. Instead of
forcing the CSV code to do its own duplication-allowance by appending
to keys (like the distsqlrun/diskRowContainer), we can more easily
do it here in the MVCC timestamp.

This uncovered a bug in the distributed CSV import that failed
to detect duplicate PK and unique index keys. Using the common
RocksDBStore now makes the code the same across both CSV
implementations.

Fixes #17707

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17712)
<!-- Reviewable:end -->

bdarnell commented:
SQL commands shouldn't be able to write to arbitrary places on the server's filesystem. We could make this a setting, but we shouldn't even require that the appropriate temp directory be the same across all the nodes in the cluster. I think that if we do anything beyond telling people to set `TMPDIR` appropriately, it should be a command-line flag. 

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-:-KrlX3oNqJLpLsihpfof:bgcuz83)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Good point. I've removed the change to the SQL command, which will continue to use the OS temp dir for now.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-:-KrlZMJpevXhLmWV7SQy:b9n3l1i)*
<!-- Sent from Reviewable.io -->

dt commented:
for a non-server cli process, I think we can just suggest `TMPDIR=/some/path cockroach load ...` ? 

@bdarnell fwiw, one of the allowed storage schemes for backup and restore is "node's local filesystem", so we already have sql access to arbitrary paths (permissions notwithstanding) -- indeed, that was the motivation for  making those statements root-only rather than just requiring relevant read or create perms on the db being backed up or imported into.

bdarnell commented:
Didn't we have some restriction on the node-local paths for backup/restore? If not, we should (e.g. the paths should be not be allowed to be absolute, they must be relative to some path specified on the server's command line)

mjibson commented:
@dt Yes, we can indeed suggest setting TMPDIR like that (we fixed a user problem that way). However there are questions about how to do that in a way that users will see that recommendation. Using an explicit flag like this makes it super clear how to change it, since our other option is to stick it in the general help text for the function which is some paragraphs long and may never be read. Maybe @robert-s-lee has some more thoughts. I'm happy with whichever solution as long as it works well for users.

dt commented:
@mjibson the distributed import (which I'd expect is what our enterprise users will use) doesn't use this, so we're adding it to the common options and plumbing it though just for the single-process case, but that can already just get it from TMPDIR. If that's just because we think someone will look at the docs just long enough to read the args but then stop before reading the text, I'm not sure that's worth it.

@bdarnell we aren't currently doing anything to restrict `nodelocal://` paths (other than os fs perms). We recommend in docs using the `nodelocal` scheme to point to directories on an NFS or other network fs mount point, and anecdotally, we've seen it used to import a downloaded backup from $HOME/Downloads or whatever.

robert-s-lee commented:
Because of potential space requirements, explicitly specified option would work better.

mjibson commented:
This PR has been completely rewritten. PTAL @dt 

@asubiotto can you review the RocksDBMap changes?

asubiotto commented on pkg/storage/engine/disk_map.go:
```diff
@@ -158,8 +160,25 @@ func generateTempStorageID() uint64 {
+// by a unique prefix. Unlike NewRocksDBMap, Puts with identical keys will
+// write multiple entries (instead of overwriting previous entries) that
+// will be returned during iteration.
+func NewRocksDBStore(e Engine) *RocksDBMap {
```
Why not just make `NewRocksDBMap` take in an `allowDuplicates` bool and have the caller specify this rather than choose between `NewRocksDBMap` and `NewRocksDBStore`? I think it would be much clearer.

asubiotto commented on pkg/storage/engine/disk_map.go:
```diff
@@ -176,13 +195,28 @@ func (r *RocksDBMap) makeKey(k []byte) MVCCKey {
+// to assign a unique rowID so duplicate keys don't overwrite each other.
+func (r *RocksDBMap) makePutKey(k []byte) MVCCKey {
+	mvccKey := r.makeKey(k)
+	if r.allowDuplicates {
```
Why not just do this in `makeKey` and get rid of `makePutKey`?

asubiotto commented on pkg/storage/engine/disk_map.go:
```diff
@@ -176,13 +195,28 @@ func (r *RocksDBMap) makeKey(k []byte) MVCCKey {
+func (r *RocksDBMap) makePutKey(k []byte) MVCCKey {
+	mvccKey := r.makeKey(k)
+	if r.allowDuplicates {
+		r.rowID++
```
I would change `rowID` to `keyID` since we're dealing with keys, not rows.

asubiotto commented on pkg/storage/engine/disk_map_test.go:
```diff
@@ -220,6 +220,84 @@ func TestRocksDBMapSandbox(t *testing.T) {
+	}
+
+	for ti, tc := range tests {
+		t.Run(fmt.Sprint(ti), func(t *testing.T) {
```
It would be nice to have a more descriptive test name. Maybe add this to the test case struct (i.e. "AllowDuplicates=true").

asubiotto reviewed (commented):
Thanks for the changes, I had implemented this at a higher level (i.e. in `diskRowContainer`) but it makes sense to have it at a lower level if other users need this too. Do you mind changing `diskRowContainer` to remove its implementation of allowing duplicates to use this new functionality?

mjibson commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/engine/disk_map.go, line 171 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Ksjiu4R5uVSuYfT97MK-r1-171:-KsnsM58lWY7eZi-sReY:b7txls5) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map.go#L171)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why not just make `NewRocksDBMap` take in an `allowDuplicates` bool and have the caller specify this rather than choose between `NewRocksDBMap` and `NewRocksDBStore`? I think it would be much clearer.
</blockquote></details>

Because I remember reading https://twitter.com/ID_AA_Carmack/status/367485612627984385 many years ago and decided to apply the wisdom: don't put booleans in constructors. If it's doing something different, it's ok for it to have a different name.

---

*[pkg/storage/engine/disk_map.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Ksjiu4R5uVSuYfT97MK-r1-203:-Ksnt20ztXM3NTHeI7DG:bjz93s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map.go#L203)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why not just do this in `makeKey` and get rid of `makePutKey`?
</blockquote></details>

Because during Scan we need to NOT add the timestamp, otherwise we could be after the key that was scanned to.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Sadly, we cannot just remove this functionality from the diskRowContainer. The HashJoiner depends on the diskRowContainer doing its own rowID appending, but also depends on the underlying rocksdbmap to not do the rowID stuff (see https://github.com/cockroachdb/cockroach/blob/20bd2bc2d1854d328b8ceac74fe22609b159feb3/pkg/sql/distsqlrun/hash_row_container.go#L493 where the hash row container needs to overwrite an existing record).

madelynnblue committed:
sqlccl: specify temp directory during local CSV transforms

Change the local CSV transform to take a RocksDB engine instead of
a directory, which allows for correctly supporting the CLI (OS temp)
and server (temp storage) use cases.

This adds a new flag to the `load csv` subcommand: tempdir, which
overrides the OS temp dir used for intermediate storage. Notably the
default is listed in the help text, making it explicit to the user
where the data will be stored if no option is passed.

For the SQL invocation that uses the local transform, pass down the
server's temp storage instead of using the OS temp dir.

This adds a new function that creates a RocksDB store instead of a
map. That is, it allows for duplicate keys instead of overwriting
them. This functionality is needed for CSV writing, which requires
that duplicate keys are kept so it can detect primary key or unique
index violations. The CSV code was previously doing that itself, but
has been moved to use the common temp storage location. Instead of
forcing the CSV code to do its own duplication-allowance by appending
to keys (like the distsqlrun/diskRowContainer), we can more easily
do it here in the MVCC timestamp.

This uncovered a bug in the distributed CSV import that failed
to detect duplicate PK and unique index keys. Using the common
RocksDBStore now makes the code the same across both CSV
implementations.

Fixes #17707

mjibson commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/engine/disk_map.go, line 204 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Ksjiu4R5uVSuYfT97MK-r1-204:-Kso5rBCmrab1GO6l2pX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map.go#L204)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would change `rowID` to `keyID` since we're dealing with keys, not rows.
</blockquote></details>

Done.

---

*[pkg/storage/engine/disk_map_test.go, line 263 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Ksjiu4R5uVSuYfT97ML-r1-263:-Kso5s3gubIsCXi-8pNO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map_test.go#L263)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It would be nice to have a more descriptive test name. Maybe add this to the test case struct (i.e. "AllowDuplicates=true").
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/engine/disk_map.go, line 171 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Ksjiu4R5uVSuYfT97MK-r1-171:-Kso4RHgUPYFD6IfaDbb:b88k24f) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map.go#L171)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Because I remember reading https://twitter.com/ID_AA_Carmack/status/367485612627984385 many years ago and decided to apply the wisdom: don't put booleans in constructors. If it's doing something different, it's ok for it to have a different name.
</blockquote></details>

Sounds good, makes sense. The only thing is that the difference between `RocksDBMap` and `RocksDBStore` is not immediately apparent to me. Consider renaming them to `RocksDBMapWithoutDuplicates` and `RocksDBMapWithDuplicates`? Or something else.

---

*[pkg/storage/engine/disk_map.go, line 201 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Kso5MX2k2dbFAws3L2-:-Kso5MX2k2dbFAws3L20:b-t9v5g4) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map.go#L201)):*
> ```Go
> // except it respects allowDuplicates, which uses the MVCC timestamp field
> // to assign a unique rowID so duplicate keys don't overwrite each other.
> func (r *RocksDBMap) makePutKey(k []byte) MVCCKey {
> ```

Consider renaming this to `makeKeyWithTimestamp` or `makeKeyWithID` or something. These kind of names make more sense to me than `makePutKey` because it describes what the method does rather than what it should be used for.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/engine/disk_map.go, line 171 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Ksjiu4R5uVSuYfT97MK-r1-171:-KsoESOreHML3CqtuPEM:bl5k4eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map.go#L171)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Sounds good, makes sense. The only thing is that the difference between `RocksDBMap` and `RocksDBStore` is not immediately apparent to me. Consider renaming them to `RocksDBMapWithoutDuplicates` and `RocksDBMapWithDuplicates`? Or something else.
</blockquote></details>

How about RocksDBMultiMap, since that's kinda what it is: https://en.wikipedia.org/wiki/Multimap

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/engine/disk_map.go, line 171 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-Ksjiu4R5uVSuYfT97MK-r1-171:-KsoFVggUnKBwucy-UcE:b-ul8jrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d25ec2eae4aea615a0a71ad662e4fd20bb5201/pkg/storage/engine/disk_map.go#L171)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

How about RocksDBMultiMap, since that's kinda what it is: https://en.wikipedia.org/wiki/Multimap
</blockquote></details>

Good idea

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712)*
<!-- Sent from Reviewable.io -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17712#-:-Ksp5CSZTlwVSNYbWlkJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->
